zkevm-circuits/src/util/cell_manager.rs

```diff
@@ -1,8 +1,269 @@
+    pub(crate) fn storage_for_phase(phase: u8) -> CellType {
+        match phase {
+            0 => CellType::StoragePhase1,
+            1 => CellType::StoragePhase2,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the storage cell of the expression
+    pub(crate) fn storage_for_expr<F: Field>(expr: &Expression<F>) -> CellType {
+        Self::storage_for_phase(Self::expr_phase::<F>(expr))
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Cell
+pub(crate) struct Cell<F> {
+    pub(crate) expression: Expression<F>,
+    pub(crate) column_expression: Expression<F>,
+    pub(crate) column: Column<Advice>,
+    pub(crate) column_idx: usize,
+    pub(crate) rotation: usize,
+
+    // DEPRECATED: the same value as column_idx
+    pub(crate) cell_column_index: usize,
+}
+
+impl<F: Field> Cell<F> {
+    pub fn new(
+        meta: &mut VirtualCells<F>,
+        column: Column<Advice>,
+        column_idx: usize,
+        rotation: usize,
+    ) -> Cell<F> {
+        Cell {
+            expression: meta.query_advice(column, Rotation(rotation as i32)),
+            column_expression: meta.query_advice(column, Rotation::cur()),
+            column,
+            column_idx,
+            rotation,
+            cell_column_index: column_idx,
+        }
+    }
+}
+
+impl<F: Field> Cell<F> {
+    pub fn new_from_cs(
+        meta: &mut ConstraintSystem<F>,
+        column: Column<Advice>,
+        column_idx: usize,
+        rotation: usize,
+    ) -> Cell<F> {
+        query_expression(meta, |meta| Cell::new(meta, column, column_idx, rotation))
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        value: Value<F>,
+    ) -> Result<AssignedCell<F, F>, Error> {
+        region.assign_advice(
+            || {
+                format!(
+                    "Cell column: {:?} and rotation: {}",
+                    self.column, self.rotation
+                )
+            },
+            self.column,
+            offset + self.rotation,
+            || value,
+        )
+    }
+}
+
+impl<F: Field> Expr<F> for Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+impl<F: Field> Expr<F> for &Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+#[derive(Debug, Clone)]
+/// CellColumn
+pub(crate) struct CellColumn {
+    pub advice: Column<Advice>,
+    pub cell_type: CellType,
+    pub idx: usize,
+}
+
+impl CellColumn {
+    pub fn new(advice: Column<Advice>, cell_type: CellType, idx: usize) -> CellColumn {
+        CellColumn {
+            advice,
+            cell_type,
+            idx,
+        }
+    }
+
+    pub fn expr<F: Field>(&self, meta: &mut ConstraintSystem<F>) -> Expression<F> {
+        query_expression(meta, |meta| meta.query_advice(self.advice, Rotation::cur()))
+    }
+}
+
+/// CellManagerStrategy
+pub(crate) trait CellManagerStrategy {
```

Comment:
This trait should work for all circuits right?  But I see that `query_cell` uses an argument `CellManagerColumns`, which depends on `CellType`, which currently is specific to the EVM Circuit.  Do you think that `CellType` could work with the other circuits?  If not, do you have any idea on how to address this?

Reply:
So, the idea is that the keccak circuit will have a default CellType, but in general the CellType concept is necessary to make it general.